workflow/controller/entrypoint/container_registery_index.go

```diff
@@ -0,0 +1,53 @@
+package entrypoint
+
+import (
+	"context"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type containerRegistryIndex struct {
+	kubernetesClient kubernetes.Interface
+}
+
+func (i *containerRegistryIndex) Lookup(ctx context.Context, image string, options Options) (*Image, error) {
+	kc, err := k8schain.New(ctx, i.kubernetesClient, k8schain.Options{
+		Namespace:          options.Namespace,
+		ServiceAccountName: options.ServiceAccountName,
+		ImagePullSecrets:   imagePullSecretNames(options.ImagePullSecrets),
+	})
+	if err != nil {
+		return nil, err
+	}
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return nil, err
+	}
+	img, err := remote.Image(ref, remote.WithAuthFromKeychain(kc))
+	if err != nil {
+		return nil, err
+	}
+	f, err := img.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	cmd := &Image{}
+	if len(f.Config.Entrypoint) > 0 {
+		cmd.Command = f.Config.Entrypoint
+	} else {
+		cmd.Command = f.Config.Cmd
+	}
+	return cmd, nil
```

Comment:
I left this comment [_elsewhere_](https://github.com/alexec/argo-workflows/commit/4bff0bc88d15b05d0c31147f3eb901c7dc8724e8) but seems that is not this MR...

This could be surprising.  The `template.container` may already have `args` specified which would replace the `f.Config.Cmd`, but this approach would effectively append `template.container.args` to the end of `f.Config.Cmd`.  Should probably return a

```
type EntrypointAndCmd struct {
    Entrypoint []string
    Cmd []string
}
```
So that downstream (where you prepend `argoexec emissary`) you could decide if the `CMD` should be used, or is replaced by `template.container.args`.

Right?

Reply:
Also, this ignores `CMD` if there's an `ENTRYPOINT`, correct? (I could be misreading). 

For example, if a `Dockerfile` had:
```Dockerfile
ENTRYPOINT [initialization_script.sh]
CMD [do_the_thing.sh]
```

This code would effectively drop the `do_the_thing.sh`, when instead it needs to concatenate `CMD` onto `ENTRYPOINT`, assuming the `template` doesn't override either as @lucastheisen suggests.

Reply:
I struggle to know what is correct here. The fact that cmd is an array seems very odd to me. I think we may have been lumbered with some very old Docker design decision. 

This logic, I copied from Tekton.

https://github.com/tektoncd/pipeline/blob/62c9b0d861bd2ae01263dfa67ba41fb0fa2bc9a9/pkg/pod/entrypoint_lookup_impl.go#L163

Reply:
I would agree with the statement:

> behaviour must be exactly the same as a Dockerfile

but that does require deviation from what Tekton is doing.  They do indeed appear to be using `ENTRYPOINT || CMD` style logic which is definitely not the _same as a Dockerfile_.  It should be relatively simple to propagate the tuple of entrypoint and cmd to the consumer (though it may have impact on your cache code that i didn't think through).  But at least then you can faithfully reproduce the ENTRYPOINT / CMD behavior.

Reply:
Case 1 - nothing user specified:

```
(image_entrypoint, image_cmd, image_args, nil, nil) -> (concat(image_entrypoint, image_cmd), image_args)
```

Case 2 - fully user specified:

```
(image_entrypoint, image_cmd, image_args, user_cmd, user_args) -> (user_cmd, user_args), image_args)
```

Case 3 - no user cmd

```
(image_entrypoint, image_cmd, image_args, nil, user_args) -> (image_entrypoint, image_args)
```

Reply:
https://stackoverflow.com/questions/44316361/difference-between-docker-entrypoint-and-kubernetes-container-spec-command
https://sahansera.dev/closer-look-at-kubernetes-pod-commands-args/

Reply:
```
If you do not supply command or args for a Container, the defaults defined in the Docker image are used.
If you supply only args for a Container, the default Entrypoint defined in the Docker image is run with the args that you supplied.
If you supply a command for a Container, only the supplied command is used. The default EntryPoint and the default Cmd defined in the Docker image are ignored. Your command is run with the args supplied (or no args if none supplied).
```